Array Documentation

slayful opened the pull request:
https://github.com/ponylang/ponyc/issues/1828

SeanTAllen commented on packages/builtin/array.pony:
````diff
@@ -1,6 +1,139 @@
+      second_element_is_wombat(["star"; "wombat"]) == true
+  ```
+
+  # Memory allocation
````
this should be ## not #

SeanTAllen commented on packages/builtin/array.pony:
```diff
@@ -1,6 +1,139 @@
   """
   Contiguous, resizable memory to store elements of type A.
+
+  # Usage
```
this should also be ## not #

mfelsche commented on packages/builtin/array.pony:
````diff
@@ -1,6 +1,139 @@
+      wombats.space() == 8 or wombats.space() == 16
+  ```
+
+  It's possible to create an empty array with at some memory allocated.
````
typo: `It is possible to create an empty Array with some memory allocated.`

SeanTAllen commented on packages/builtin/array.pony:
````diff
@@ -1,6 +1,139 @@
+    let array: Array[String] = ["dog"; "cat"; "wombat"]
+  ```
+
+  Semicolon is an expression, so this is a multiline equivalent.
````
the multiline equiv here can be removed. that's not about array, its about a language feature.

mfelsche commented on packages/builtin/array.pony:
```diff
@@ -1,6 +1,139 @@
+
+  # Memory allocation
+  Array allocates contiguous memory. It always allocates at least enough memory
+  space to hold all of it's elements. Space is the number of elements the array
```
typo: `its elements`

mfelsche commented on packages/builtin/array.pony:
```diff
@@ -1,6 +1,139 @@
+  can hold without allocating more memory. The exact amout of memory needed for
+  one unit of space depends on the type of elements the array can hold.
+
+  When creating an array or adding more elements the Array will allocate
```
typo: use uppercase  `Array`

SeanTAllen commented on packages/builtin/array.pony:
```diff
@@ -1,6 +1,139 @@
+  Array allocates contiguous memory. It always allocates at least enough memory
+  space to hold all of it's elements. Space is the number of elements the array
+  can hold without allocating more memory. The exact amout of memory needed for
+  one unit of space depends on the type of elements the array can hold.
```
I believe the idea here is that an Array[U64] with size of 6 will take more memory than an Array[U8] of the same size.

I think that point could be made more clear with some a couple of concrete examples.

mfelsche commented on packages/builtin/array.pony:
```diff
@@ -1,6 +1,139 @@
+  one unit of space depends on the type of elements the array can hold.
+
+  When creating an array or adding more elements the Array will allocate
+  a next power of two of space, but no less than 8 of space. Empty array
```
typo: `An empty Array...`

SeanTAllen commented on packages/builtin/array.pony:
````diff
@@ -1,6 +1,139 @@
+  ```pony
+      // create an array of 8 wombats
+      let wombats = Array[String].init("wombat", 8)
+      wombats.size() == 8
````
i think these comparisons should be comments.

mfelsche commented on packages/builtin/array.pony:
````diff
@@ -1,6 +1,139 @@
+      empty_array.space() == 16
+  ```
+
+  # More examples
````
Most of those examples should be placed in the respective methods they describe.
The class documentation here should focus on an overview of what is available in this class and describe semantics, behavior and usage of this class in general.

Thus is would remove this section here and move some of the content.

slayful commented:
Thanks for the feedback @SeanTAllen, @mfelsche. :) Listening to the dev sync helped provide some more context too.

I believe I've addressed most of it. I don't believe moving any of the examples to the methods' docs is required - they seem pretty self explanatory. I've replaced 'more examples' with an overview with most useful methods in the beginning.

Ready for second CR round.

mfelsche commented on packages/builtin/array.pony:
````diff
@@ -1,6 +1,78 @@
+  ```pony
+    let array: Array[String] = ["dog"; "cat"; "wombat"]
+  ```
+
````
maybe also show an equivalent example using an explicit constructor.

mfelsche commented on packages/builtin/array.pony:
````diff
@@ -1,6 +1,78 @@
+  ```
+
+  Accessing elements can be done via the `apply(i: USize): this->A ?` method.
+  The provided index might be out of bounds so `apply` (like many other Array's
````
I'd remove the part in parentheses.

mfelsche commented on packages/builtin/array.pony:
````diff
@@ -1,6 +1,78 @@
+  ```
+
+  Adding and removing elements to and from the end of the Array can be done via
+  `push` and `pop` methods, which add and remove elements to the end of the
````
The part after the comma seems duplicated and can be removed.

mfelsche commented on packages/builtin/array.pony:
```diff
@@ -1,6 +1,78 @@
+  which alter the Array at an arbitrary index, moving elements left (when
+  deleting) or right (when inserting) as necessary.
+
+  Iterating over can be done using the `values` method.
```
Something seems missing here. Suggestion: `Iterating over the elements of an Array`.

This particular feature of Array definitely needs some more explanation, as it returns aliases on the elemens as the array sees them (taking viewpoint adaptation into account), which is a complicated topic for some cases. But this can be part of another PR.

mfelsche commented on packages/builtin/array.pony:
````diff
@@ -1,6 +1,78 @@
+    end
+  ```
+
+  You could also treat the array as a FIFO queue using `pop()` method.
````
maybe move this small remark and the example up where `pop()` and `push() are described.

mfelsche commented on packages/builtin/array.pony:
```diff
@@ -1,6 +1,78 @@
+  of 6 will take more memory than an Array[U8] of the same size.
+
+  When creating an Array or adding more elements the Array will allocate
+  a next power of two of space, but no less than 8 of space. An empty Array
```
maybe a more clear formulation: `the Array will calculate the next power of 2 of the requested number of elements and allocate that much space, with a lower bound of space for 8 elements.`

mfelsche commented on packages/builtin/array.pony:
```diff
@@ -1,6 +1,78 @@
+  a next power of two of space, but no less than 8 of space. An empty Array
+  is an exception and can be of 0 space.
+
+  Here's a few examples of the space allocated when initialising a table with
```
typo: `when initialising an Array with ...`

slayful commented:
Thank you @mfelsche, addressed your feedback.

I haven't got around to address this though, I'm not sure I can explain this clearly enough, as I don't understand it completely.
> This particular feature of Array definitely needs some more explanation, as it returns aliases on the elemens as the array sees them (taking viewpoint adaptation into account), which is a complicated topic for some cases. But this can be part of another PR.

mfelsche commented on packages/builtin/array.pony:
```diff
@@ -1,6 +1,86 @@
+
+  Adding and removing elements to and from the end of the Array can be done via
+  `push` and `pop` methods. You could treat the array as a FIFO queue using
+  those method.
```
typo: methods

mfelsche commented on packages/builtin/array.pony:
````diff
@@ -1,6 +1,86 @@
+  those method.
+  ```pony
+    while (array.size() > 0) do
+      let elem = array.pop()
````
`pop` is actually partial, so we need a `?` here.

slayful commented:
Thanks @mfelsche - addressed your feedback.

mfelsche commented:
Good to go!

SeanTAllen commented:
Thanks @slayful. I squashed into a single commit and merged.
